### PR TITLE
Topological generations

### PR DESCRIPTION
**Pull request by 16arpi**
Pull request to add a new feature to the actual topological sort. It updates the topological sort algorithm in a way to get the topological generation at each iteration. It adds two new methods : 

- `topologicalGenerations(graph)`
- `forEachTopologicalGeneration(graph)`

The algorithm used here comes from the `networkx` python library.

### CODE REVIEW

File: src/dag/test.js [hunk 1]
```diff
@@ -17,6 +17,8 @@ const {
   topologicalSort,
   forEachNodeInTopologicalOrder
 } = require('./topological-sort.js');
+const { topologicalGenerations } = require('./topological-sort.js');
```
Comment (Yomguithereal) [inline src/dag/test.js:20]: Pourquoi tu ne mets pas ces imports dans le require précédent?

File: src/dag/test.js [hunk 2]
```diff
@@ -268,5 +270,91 @@ describe('graphology-dag', function () {
...
+      graph.mergeEdge(2, 3);
+      graph.mergeEdge(3, 4);
+      graph.mergeEdge(4, 5);
+      graph.mergeEdge(5, 6);
+
+      assert.strictEqual(graph.order, 6);
+      assert.strictEqual(graph.size, 5);
+
+      const generations = topologicalGenerations(graph);
+      const generations_set = generations.map(gen => {
```
Comment (Yomguithereal) [inline src/dag/test.js:287]: Comme dit la CI, les noms de variables doivent être en camelCase, pas en snake_case, on est pas en python ici :)

Comment (Yomguithereal) [reply src/dag/test.js:287]: Note que dans le projet tu peux lancer `npm run lint` pour voir ce genre d'erreur. Et à la racine tu peux lancer `npm run prettier` pour formatter le code un peu comme quand on utilise `black` en python.

File: src/dag/test.js [hunk 3]
```diff
@@ -268,5 +270,91 @@ describe('graphology-dag', function () {
...
+        return new Set(gen)
+      })
+
+      console.log(generations_set);
```
Comment (Yomguithereal) [inline src/dag/test.js:291]: Il ne faut pas laisser de log en prod.

File: src/types/index.d.ts [hunk 4]
```diff
@@ -98,6 +98,12 @@ type NodeIterationCallback<NodeAttributes extends Attributes = Attributes> = (
   attributes: NodeAttributes
 ) => void;
 
+type NodeIterationCallbackWithGenerations<NodeAttributes extends Attributes = Attributes> = (
```
Comment (Yomguithereal) [inline src/types/index.d.ts:101]: NodeIterationCallbackWithGeneration au singulier.

File: src/dag/topological-sort.d.ts [hunk 5]
```diff
@@ -7,3 +7,10 @@ export function forEachNodeInTopologicalOrder<
   callback: NodeIterationCallback<NodeAttributes>
```
Comment (Yomguithereal) [inline src/dag/topological-sort.d.ts:7]: Tu dois changer le type de cette callback pour NodeIterationCallbackWithGeneration.

File: src/dag/topological-sort.d.ts [hunk 6]
```diff
@@ -7,3 +7,10 @@ export function forEachNodeInTopologicalOrder<
...
 ): void;
 export function topologicalSort(graph: Graph): Array<string>;
+export function forEachTopologicalGeneration<
+  NodeAttributes extends Attributes = Attributes
+>(
+  graph: Graph<NodeAttributes>,
+  callback: NodeIterationCallbackWithGenerations<NodeAttributes>
```
Comment (Yomguithereal) [inline src/dag/topological-sort.d.ts:14]: Ce n'est pas le bon type de callback. Là on appelle la fonction avec une liste de strings non?

File: src/dag/topological-sort.d.ts [hunk 7]
```diff
@@ -7,3 +7,10 @@ export function forEachNodeInTopologicalOrder<
...
+): void;
+export function topologicalGenerations(graph: Graph): string[];
```
Comment (Yomguithereal) [inline src/dag/topological-sort.d.ts:16]: Le reste du code favorise la notation `Array<string>` plutôt (ya pas d'argument en faveur de l'un ou l'autre, juste aligner la convention).

File: src/dag/topological-sort.js [hunk 8]
```diff
@@ -59,11 +59,13 @@ function forEachNodeInTopologicalOrder(graph, callback) {
...
 
   while (queue.size !== 0) {
-    const [node, attr] = queue.shift();
+    const [node, attr, gen] = queue.shift();
 
-    callback(node, attr);
+    callback(node, attr, gen);
 
-    graph.forEachOutNeighbor(node, neighborCallback);
+    graph.forEachOutNeighbor(node, (node, attr) => {
```
Comment (Yomguithereal) [inline src/dag/topological-sort.js:66]: Tu viens de faire ce qu'on avait dit qu'il ne fallait pas faire :). A savoir créer une fonction à chaque itération de la boucle `while`. Ca et tu shadow la variable `node` de manière assez violente là.

Comment (Yomguithereal) [reply src/dag/topological-sort.js:66]: Il te faut une variable d'état `currentGeneration` que tu set avant d'itérer sur les voisins, que tu utilises dans `neighborCallback` qui n'a plus besoin de l'argument `gen` et l'affaire est dans le sac.

File: src/dag/topological-sort.js [hunk 9]
```diff
@@ -88,8 +90,47 @@ function topologicalSort(graph) {
...
+      if (gen > lastGenLevel) {
+        callback(lastGen);
+        lastGenLevel = gen;
+        lastGen = [];
+      }
+
+      lastGen.push(node);
+    });
+
+    callback(lastGen);
```
Comment (Yomguithereal) [inline src/dag/topological-sort.js:112]: Je réfléchis à haute voix mais il est pas possible à ce stade que `lastGen` soit vide, dans un edge case? Genre il se passe quoi si il n'y a qu'une seule génération ou si il n'y a pas de nodes dans le graph (il faut probablement ajouter des tests pour ces cas d'ailleurs)?

**Review by Yomguithereal — CHANGES_REQUESTED** [review]

File: src/dag/topological-sort.d.ts (hunk 6 above)
Comment (16arpi) [reply src/dag/topological-sort.d.ts:14]: Oui totalement

File: src/dag/topological-sort.d.ts (hunk 5 above)
Comment (16arpi) [reply src/dag/topological-sort.d.ts:7]: Je devrais peut-être enlever le NodeIterationCallbackWithGeneration vu que le NodeIterationCallback renvoie déjà la génération ? Comme ça je rajouter `generation: number` à NodeIterationCallback.

File: src/dag/topological-sort.js (hunk 9 above)
Comment (16arpi) [reply src/dag/topological-sort.js:112]: J'ai testé les 2 éventualités. Quand il n'y a que des nœuds et pas d'arc, ça retourne bien une seule génération. Quant au graphe vide, ça renvoie une erreur. Peut-être ajouter à `forEachNodeInTopologicalOrder()` une exception si le graphe est vide ?
